influxql/functions.go

```diff
@@ -757,3 +757,137 @@ func (r *FloatHoltWintersReducer) constrain(x []float64) {
 		x[3] = 0
 	}
 }
+
+// FloatIntegralReducer calculates the time-integral of the aggregated points.
+type FloatIntegralReducer struct {
+	groupByTime bool
+	count       uint32
+	interval    Interval
+	sum         float64
+	prev        FloatPoint
+}
+
+// NewFloatIntegralReducer creates a new FloatIntegralReducer.
+func NewFloatIntegralReducer(interval Interval, groupByTime bool) *FloatIntegralReducer {
+	return &FloatIntegralReducer{
+		groupByTime: groupByTime,
+		interval:    interval,
+		prev:        FloatPoint{Nil: true},
+	}
+}
+
+// AggregateFloat aggregates a point into the reducer.
+func (r *FloatIntegralReducer) AggregateFloat(p *FloatPoint) {
+
+	// If this is the first point, just save it
+	if r.prev.Nil {
+		r.prev = *p
+		r.count++
+		return
+	}
+
+	// If this point has the same timestamp as the previous one,
+	// use it but don't add anything to the integral.  Effectively
+	// this means that we allow the curve we are integrating
+	// to have step changes in it, but who knows whether the points
+	// actually arrive in any particular order...?
+	if r.prev.Time == p.Time {
```

Comment:
This is a potential error location. Points aren't guaranteed to be ordered. There are two ways to resolve this. Either have integral request the points be ordered from the storage engine (see `median()` in `influxql/select.go`) or by having this function operate on the entire slice instead (see `median()` again).